### DISCUSSION TITLE
OTA and Factory partition

### DISCUSSION THREAD

**Discussion by MRIIOT** [micropython/micropython#13132] [ESP32]
How come all the OTA partition examples omit the `factory` partition?  

If I wanted to rollback to factory for some reason, I would want to do something like this:

```
fp = Partition.find(label='factory')
fp.set_boot()
machine.reset()
```

**Comment by glenn20** [comment]
The default 4MB flash size on the original ESP32 devices only provided enough space for 2 firmware partitions that could fit a typical micropython application image (~1.5MB), so there is not enough room for a fallback factory partition in addition to the (minimum 2) OTA partitions.

If you have more flash storage, you can add a factory partition as well if you want (eg. see https://github.com/glenn20/mp-image-tool-esp32), though its not clear how this provides any advantage over 2 (or 3) OTA partitions.

**Reply by glenn20** [reply]
Oh and automatic rollback support is included in micropython. See https://github.com/glenn20/micropython-esp32-ota for an example of usage.

**Reply by MRIIOT** [reply]
Thanks @glenn20.  I was thinking to have a hard fallback to factory if OTA goes really wrong, but I have not convinced myself that I need it because with rollback cancel I have not been able to make an OTA upgrade fail.  We do store a json config file that might need to be upgraded during OTA if we add new functionality but I think I have that figured out as well. I think upgrading the config file was the sticking point, if the fw and config ever got out of sync then we would brick the esp and might not be able to recover unless we went back to factory fw and config.

Here is a diagram for my approach  https://github.com/SharcTech/sharc-support/blob/main/firmware/sharc_ota_diagram.pdf

**Reply by MRIIOT** [reply]
When I build a firmware and end up with micropython.bin, why do I have to use otatool.py to read from device in order to make that bin file usable for OTA?

**Reply by glenn20** [reply]
The default micropython bin files include the bootloader, the partition table and the `application image`. When you perform an ota update, you only want to write a new `application image` into the OTA partition and not overwrite the bootloader and partition table. I am guessing `otatool.py` automatically extracts the `application image` from the firmware file. Note that the `.app-bin` files on https://micropython.org/download/ESP32_GENERIC/ are `application image` files that can be used for OTA updates.

My https://github.com/glenn20/mp-image-tool-esp32 tool can also extract these `application image` files from a regular micropython firmware file.

**Reply by glenn20** [reply]
Whoops... I just realised you may have been referring to the `application image` when you mentioned `micropython.bin`. If that is the case, I have no idea why you need to use esptool.py. I don't use it and my ota updates work just fine ;).

**Reply by MRIIOT** [reply]
Yea the build produces bootloader.bin, partition-table.bin, ota_data_initial.bin, and micropython.bin.  I'll have to use your image tool to take a deeper look.

I'd love to chat if you ever have time.  cmisztur@mriiot.com

**Reply by MRIIOT** [reply]
```
mp-image-tool-esp32 % ls ../builds/v105/3ddecbe 
bootloader.bin          build.txt               micropython.bin         ota_data_initial.bin    partition-table.bin     sdkconfig.combined

mp-image-tool-esp32 % ./mp-image-tool-esp32 ../builds/v105/3ddecbe/micropython.bin
Opening image file: ../builds/v105/3ddecbe/micropython.bin... 
Chip type: esp32 
Flash size: 16MB 
PartitionError: No partition table found.
```

**Reply by glenn20** [reply]
`mp-image-tool-esp32` works with combined firmware images (or serial attached micropython boards) which aso include a partition table and a bootloader (`firmware.bin` if you are building your own micropython images). In this case, the `micropython.bin` is an `application image` which can be written directly to an **ota** partition.
